apps/rush-lib/src/logic/base/BaseShrinkwrapFile.ts

```diff
@@ -148,11 +120,13 @@ export abstract class BaseShrinkwrapFile {
    */
   public abstract getWorkspaceKeyByPath(workspaceRoot: string, projectFolder: string): string;
 
-  /** @virtual */
-  protected abstract getWorkspaceDependencyVersion(
-    dependencySpecifier: DependencySpecifier,
-    workspaceKey: string
-  ): DependencySpecifier | undefined;
+  /**
+   * Returns whether or not the workspace specified by the shrinkwrap matches the state of
+   * a given package.json. Returns true if any dependencies are not aligned with the shrinkwrap.
+   *
+   * @virtual
+   */
+  public abstract isWorkspaceProjectModified(workspaceKey: string, packageJson: PackageJsonEditor): boolean;
```

Comment:
What happens if the repo isn't using workspaces?

Reply:
If it isn't using workspaces, it uses the old logic. It's not ideal, of course, but the changes I made won't be usable in non-workspace world due to the lack of `importers` in the pnpm-lock.yaml ... we'd be probably better off with the existing logic over there.